core/modules/util/EventThread.cc

```diff
@@ -46,40 +46,98 @@ namespace util {
 void EventThread::handleCmds() {
     startup();
     while(_loop) {
-        Command::Ptr cmd = _q->getCmd();
-        if (cmd != nullptr) {
-            _q->commandStart(cmd);
-            cmd->action(this);
-            _q->commandFinish(cmd);
+        _cmd = _q->getCmd();
+        _commandFinishCalled = false;
+        _currentCommand = _cmd.get();
+        if (_cmd != nullptr) {
+            _q->commandStart(_cmd);
+            specialActions(_cmd);
+            _cmd->runAction(this);
+            callCommandFinish(_cmd);
             // Reset _func in case it has a captured Command::Ptr,
             // which would keep it alive indefinitely.
-            cmd->resetFunc();
+            _cmd->resetFunc();
         }
+        _cmd.reset();
+        _currentCommand = nullptr;
     }
     finishup();
 }
 
+
+/// Ensure that commandFinish is only called once.
+void EventThread::callCommandFinish(Command::Ptr const& cmd) {
+    if (_commandFinishCalled.exchange(true) == false) {
+        _q->commandFinish(cmd);
```

Comment:
Using std::call_once uses about the same amount of code. The interesting thing is that std::call_once only considers the function called if it does not throw an exception. If the function throws an exception, std::call_once will allow it be called until it does not throw an exception, and that's not what I'm looking for here.